### DISCUSSION TITLE
Repo for a step by step tutorial

### DISCUSSION THREAD

**Discussion by danielebarell** [community/community#23151] [New to GitHub]
<p>Hi,<br>
:slight_smile:<br>
I’m an absolute git beginner.<br>
Also I’m alone using the DCVS and no collaborative activity is intended for now.</p>
<p>I need to organize a repo for some tutorials, so that the step by step progress from start to finish is clearly shown and steps are easily available for consultation.<br>
So I thought to dedicate  a branch in the tutorial repo for every step.<br>
No merging for final solution but a simple ending branch.<br>
[see sketch below]</p>
<p>Yesterday I tried to solve by reading a getting started tutorial and using some basic commands with git-bash, but I couldn’t find a way to create the structure of the repo the way I wished.</p>
<p>In particular I couldn’t create a new branch and push the files in it.</p>
<p>Any tip is welcomed!<br>
:slightly_smiling_face:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/2/f/2f0fd856e4f5ed7caaf2f4ea718f2812546b6d7f.jpeg" data-download-href="/uploads/short-url/6Iko7qEZw2Oa1GtZ4acjiR5RvVJ.jpeg?dl=1" title="repo-structure" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/23028311/181075354-54d1b07c-ea45-4106-9c37-7f3f5c39e76c.jpeg" alt="repo-structure" data-base62-sha1="6Iko7qEZw2Oa1GtZ4acjiR5RvVJ" width="690" height="388" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/2/f/2f0fd856e4f5ed7caaf2f4ea718f2812546b6d7f_2_690x388.jpeg, https://aws1.discourse-cdn.com/github/optimized/3X/2/f/2f0fd856e4f5ed7caaf2f4ea718f2812546b6d7f_2_1035x582.jpeg 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/2/f/2f0fd856e4f5ed7caaf2f4ea718f2812546b6d7f_2_1380x776.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/2/f/2f0fd856e4f5ed7caaf2f4ea718f2812546b6d7f_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">repo-structure</span><span class="informations">3542×1993 919 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by mpboom** [comment]
<p>Hi <a class="mention" href="/danielebarell">@danielebarell</a> and welcome to the Github Community!</p>
<p>While personally, I think I would prefer making a separate directory instead of a separate branch for every step of the tutorial, the structure you want to have is certainly achievable.</p>
<p>Below I’ll make use of the Git command line, but if you prefer a more visual approach - which may be easier to get started with - you can do all these things from the <a href="https://desktop.github.com/" rel="noopener nofollow ugc">GitHub Desktop</a> app as well.</p>
<ul>
<li>To create a new branch, type - for example - <strong>git branch coding1</strong>
</li>
<li>Then switch to that branch: <strong>git checkout coding1</strong>
</li>
<li>Then add the files you wish, and make commits on the branch</li>
<li>Push the branch to GitHub, if you desire: <strong>git push --set-upstream origin coding1</strong>
</li>
<li>Go back to the first step, and make more branches!</li>
</ul>
<p>So you can always create branches using <strong>git branch</strong> and switch to them using <strong>git checkout</strong>. There are plenty of Git tutorials/QA’s available if you Google for a bit, as well as some helpful people over here.</p>
<p>Let us know if that answered your question!</p>

**Comment by danielebarell** [comment]
<p>Hi <a class="mention" href="/mpboom">@mpboom</a><br>
Happy to be part of the commmunity :slight_smile: !</p>
<p>Today I’m going to follow your instrunctions, try Bash first and then GitHub Desktop.<br>
Also I’m considering your tip about a different structure in directories, but as a workflow I thought it was better to have a single source of code in local per tutorial, no more than once.</p>
<p>Indeed what I wanted to do was a series of tutorials linked toghether pointing in a module to build up a web app.<br>
In this scenario I could</p>
<ul>
<li>use a single repo for the entire web app module (rather than one per tutorial)</li>
<li>use one directory per tutorial</li>
<li>in the single directory/tutorial use one branch per coding step.</li>
</ul>
<p>This could be the optimum.<br>
In your opinion is it “<em>easily</em>” achievable?</p>

**Comment by mpboom** [comment]
<aside class="quote no-group" data-username="danielebarell" data-post="3" data-topic="160755">
<div class="title">
<div class="quote-controls"></div>
 danielebarell:</div>
<blockquote>
<ul>
<li>use one directory per tutorial</li>
<li>in the single directory/tutorial use one branch per coding step.</li>
</ul>
</blockquote>
</aside>
<p>That isn’t really possible, since a branch hoods every directory in the repo, and thus the other tutorial directories as well.</p>
<p>However, if you swap them around (like your original plan, so a branch per tutorial) then you actually can achieve this easily.</p>
<p>One benefit of this approach would be that you can easily get a side-by-side diff of the changes.</p>

**Comment by danielebarell** [comment, marked as answer]
<p>I tried and solved the problem.<br>
Great <a class="mention" href="/mpboom">@mpboom</a> !<br>
:tada:<br>
Thanks a lot!</p>
<p>Now the next problem is to discover how to avoid to push some files or directories from local to repo.<br>
For example I use an editor with its own configuration files which are no use for tutoring.</p>
<p>Maybe It’s better  to search around for a little and open another help request just in case.</p>
<p>Thanks again :smiley:</p>

**Comment by mpboom** [comment]
<p><a class="mention" href="/danielebarell">@danielebarell</a> that is what a <a href="https://docs.github.com/en/github/using-git/ignoring-files" rel="noopener nofollow ugc">.gitignore</a> is here for :wink:</p>

**Comment by danielebarell** [comment]
<p>Yes, but info/exlude file fits even better to me.<br>
:slight_smile:</p>
